File: 2022-03-March/24/solution.cpp
#include <deque>
#include <iostream>

void	maxSub(int arr[], int k, int len)	{
	int i = 0;
	std::deque<int> ind(0);
	for (i = 0; i < k; i++)	{
		while (!ind.empty() && arr[i] >= arr[ind.back()])
			ind.pop_back();
		ind.push_back(i);
	}
	std::cout << "[" << arr[ind.front()];
	for (; i < len; i++)	{
		while (!ind.empty() && (i-k+1) > ind.front())
			ind.pop_front();
		while (!ind.empty() && arr[i] >= arr[ind.front()])
			ind.pop_front();
		ind.push_back(i);
		std::cout << ", " << arr[ind.front()];
	}
	std::cout << "]" << std::endl;
}

int	main(void){
	int arr[] = {10, 5, 2, 7, 8, 7};
	maxSub(arr, 3, 6);
}

File: 2022-03-March/24/solution.py
#!/usr/bin/env python3

from collections import deque

def max_sub(arr: list[int], k: int, le: int):
	ind = deque([])
	i = 0
	while (i < k):
		while ind and arr[i] > arr[ind[-1]]:
			ind.pop()
		ind.append(i)
		i += 1
	print('[' + str(arr[ind[0]]), end='')
	while i < le:
		while ind and ind[0] < i - k + 1:
			ind.popleft()
		while ind and arr[i] > arr[ind[0]]:
			ind.popleft()
		ind.append(i)
		print(', ' + str(arr[ind[0]]), end='')
		i += 1
	print("]")

if __name__ == "__main__":
	inp = [10, 5, 2, 7, 8, 7]
	max_sub(inp, 3, len(inp))
